### PR TITLE
Add pre-commit hook to check if the current branch needs rebase

### PR DESCRIPTION
**Pull request by bernhardkaindl**
Add pre-commit hook to check if the current branch needs rebase
- GitHub runs CI on the PR branch rebased to the target branch if possible.
- To avoid hard side effects from working on the branch that is behind the target, warn about it.

### CODE REVIEW

File: .github/workflows/check_branch-needs-rebase.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    current_branch = subprocess.check_output(
+        ["git", "rev-parse", "--abbrev-ref", "HEAD"], text=True
+    ).strip()
+
+    err, remote_ref = run(
+        ["git", "rev-parse", "--abbrev-ref", f"{current_branch}@{{upstream}}"],
+    )
+    if err:
+        target = "origin/master"
+        # Get the default branch name of the remote repository
```
Comment (GeraldEV) [inline .github/workflows/check_branch-needs-rebase.py:74]: Should this comment be after the else line?

File: .github/workflows/check_branch-needs-rebase.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    """Run a command and return the output."""
+    cmd = subprocess.run(command, check=check, text=True, capture_output=True)
+    return cmd.returncode, cmd.stdout.strip()
+
+
+# TODO Rename this here and in `main`
+def check_if_local_branch_is_behind_target(current_branch, target="origin/master"):
+    """Check if the current branch is behind the target branch."""
+
+    log(f"The current branch {current_branch} does not have an upstream branch")
```
[the remark this answers is not in the document]
Comment (bernhardkaindl) [reply .github/workflows/check_branch-needs-rebase.py]: Thanks for the review! Correct!

I updated this and fixed it, added missing comments as well!

File: .github/workflows/check_branch-needs-rebase.py (hunk 1 above)
Comment (bernhardkaindl) [reply .github/workflows/check_branch-needs-rebase.py]: Thanks for the review! Correct, this was the wrong location!

Moved it and added the correct comment at that place!

**Comment by bernhardkaindl** [comment]
@GeraldEV Thanks for the review!

Thanks for hinting me at the log() call and the comment!

I updated both, fixing them and moved the comment and added missing comments as well!

File: .github/workflows/check_branch-needs-rebase.py
Comment (ashwin9390) [inline .github/workflows/check_branch-needs-rebase.py:1]: pylance throws annotation error during runtime for list and tuple ,should you consider using quotes for it ?.

**Review by ashwin9390 — APPROVED** [review]

File: .github/workflows/check_branch-needs-rebase.py
Comment (bernhardkaindl) [reply .github/workflows/check_branch-needs-rebase.py:1]: I guess you refer to:
```py
def run(command: list[str], check=False) -> tuple[int, str]:
```
This these annotations are fine for Python3 code, and VS Code/Pylance should only warn a warning about it if you selected Python 2.7 as the active Python interpreter in VS Code: Don't do that normally. It's better if you select a Python3 version as the active interpreter.

Pyright (which is the static checker in Pylance) has no issues:
```py
pyright .github/workflows/check_branch-needs-rebase.py
0 errors, 0 warnings, 0 informations 
```

**Review by GeraldEV — APPROVED** [review]

**Pull request merged**